### PR TITLE
[CDAP-18701] Fix program state transition from STARTING -> RUNNING up…

### PR DESCRIPTION
**Pull request by CuriousVini**
…on pod RUNNING State

image

image

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java [hunk 1]
```diff
@@ -75,9 +75,11 @@
   private final BatchV1Api batchV1Api;
   private final Type resourceType;
   private final V1ObjectMeta meta;
+  private final CompletableFuture<Void> started;
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java]: nit: startupTaskFuture

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java [hunk 2]
```diff
@@ -245,7 +248,11 @@ public void kill() {
 
   @Override
   public void onRunning(Runnable runnable, Executor executor) {
-    executor.execute(runnable);
+    if (resourceType.equals(V1Job.class)) {
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java:251]: Add comments why this is safe for V1Jobs.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 3]
```diff
@@ -423,7 +433,29 @@ private boolean isJobReady(V1Job job) {
...
+                                                        labelSelector, null, null, null, null, null);
+
+        for (V1Pod pod : podList.getItems()) {
+          if (pod.getStatus() != null && pod.getStatus().getPhase() != null
+            && pod.getStatus().getPhase().equalsIgnoreCase("RUNNING")) {
+            return true;
+          }
+        }
+      } catch (ApiException e) {
+        // If there is an exception while getting active pods for a job, we will use job level status.
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java:457]: Shouldn't we log this exception?

**Review by masoud-io — COMMENTED** [review]
Shouldn't we first merge this into `develop` branch?

**Comment by CuriousVini** [comment]
> Shouldn't we first merge this into `develop` branch?

Yes it should be cherrypicked to develop too. Will open that PR next

**Review by masoud-io — APPROVED** [review]

**Pull request merged**